src/org/sosy_lab/common/io/TempDir.java

```diff
@@ -0,0 +1,167 @@
+import static org.sosy_lab.common.io.TempFile.TMPDIR;
+
+import com.google.common.base.Strings;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import com.google.errorprone.annotations.Immutable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class TempDir {
+
+  private TempDir() {}
+
+  /** Create a builder for temporary directories. */
+  public static TempDirBuilder builder() {
+    return new TempDirBuilder();
+  }
+
+  public static final class TempDirBuilder {
+
+    private Path dir = TMPDIR;
+    private @Nullable String prefix;
+    private boolean deleteOnJvmExit = true;
+    private FileAttribute<?>[] fileAttributes = new FileAttribute<?>[0];
+
+    private TempDirBuilder() {}
+
+    /** The directory where the temp dir will be created, default is JVM's temp directory. */
+    @CanIgnoreReturnValue
+    public TempDirBuilder dir(Path pDir) {
+      dir = checkNotNull(pDir);
+      return this;
+    }
+
+    /** Prefix of randomly-generated directory name. */
+    @CanIgnoreReturnValue
+    public TempDirBuilder prefix(String pPrefix) {
+      prefix = checkNotNull(pPrefix);
+      return this;
+    }
+
+    /** Do not automatically delete the directory including its contents on JVM exit. */
+    @CanIgnoreReturnValue
+    public TempDirBuilder noDeleteOnJvmExit() {
+      deleteOnJvmExit = false;
+      return this;
+    }
+
+    /** Use the specified {@link FileAttribute}s for creating the directory. */
+    @CanIgnoreReturnValue
+    public TempDirBuilder fileAttributes(FileAttribute<?>... pFileAttributes) {
+      fileAttributes = pFileAttributes.clone();
+      return this;
+    }
+
+    /**
+     * Create a fresh temporary directory according to the specifications set on this builder.
+     *
+     * <p>If the temporary directory should be removed including its contents after some specific
+     * code is executed, use {@link #createDeleteOnClose()}.
+     *
+     * <p>This instance can be safely used again afterwards.
+     */
+    public Path create() throws IOException {
+      Path tempDir;
+      try {
+        tempDir = Files.createTempDirectory(dir, prefix, fileAttributes);
+      } catch (IOException e) {
+        // The message of this exception is often quite unhelpful,
+        // improve it by adding the path where we attempted to write.
+        if (e.getMessage() != null && e.getMessage().contains(dir.toString())) {
+          throw e;
+        }
+
+        String dirName = dir.resolve(prefix + "*").toString();
+        if (Strings.nullToEmpty(e.getMessage()).isEmpty()) {
+          throw new IOException(dirName, e);
+        } else {
+          throw new IOException(dirName + " (" + e.getMessage() + ")", e);
+        }
+      }
+
+      if (deleteOnJvmExit) {
```

Comment:
Hm, do you really need this feature?

AFAIU the equivalent feature for files is somewhat fragile and should be avoided if there are other alternatives. For directories, it would be worse: First, the current implementation adds on thread for each directory, which can lead to a huge number of threads and overhead. Second, deleting a directory recursively can take an arbitrary long time and block the JVM shutdown, with a debugger needed for finding out why the program seems to hang. Third, if other threads continue to run while the shutdown hook is already active (this happens if the JVM receives Ctrl+C, for example), all kinds of problems can arise when two threads try to manipulate files in the same directory.

So I would say this feature should ideally not be added, only if there is a strong argument for it (and even then, maybe not add it in a general place where it can be easily misused, but specifically for where it is needed?).

Reply:
No, it is not needed